### ISSUE TITLE
Detect MEMORY_PROTECTION

### ISSUE THREAD

**Issue by mpsido** [skycoin/hardware-wallet#45]
- [x] Memory protection query command in bootloader mode (should return RDP level)
- [x] Invert Skycoin logo in bootloader mode if in RDP level is under 2

Whenever possible prefer hardware info over `MEMORY_PROTECT` variable .

**Comment by mpsido** [comment]
@gz-c :   Can the bootloader affect the display to visually indicate whether or not memory protection is on?

**Comment by mpsido** [comment]
Use the compilation flag activating the memory protection to declare some code saying "I am not memory protected" if not defined `#ifndef MEMORY_PROTECT`...etc...

**Comment by mpsido** [comment]
Read more about memory protection here: https://www.st.com/content/ccc/resource/technical/document/application_note/89/12/c5/e2/0d/0e/45/7f/DM00186528.pdf/files/DM00186528.pdf/jcr:content/translations/en.DM00186528.pdf

**Comment by mpsido** [comment]
Solution is to create a message reception procedure in the bootloader for GetVersion message code:

This is the buffer sent to the device from the golang code if we send the GetVersion message with memoryProtect option

[getVersionMemprotect.patch.txt](https://github.com/skycoin/hardware-wallet/files/2777480/getVersionMemprotect.patch.txt)


2019/01/21 11:31:05 [[63 35 35 0 120 0 0 0 2 10 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]]

**Comment by jdknives** [comment]
@olemis is it possible to make the bootloader invert the Skycoin logo if not memory protected? Have we implemented binary labeling to indicate mem_protect status and version?

**Comment by olemis** [comment]
> is it possible to make the bootloader invert the Skycoin logo if not memory protected?

yes

> Have we implemented binary labeling to indicate mem_protect status and version?

there are versioning rules but they do not consider mem protect since every final product should be memory protected prior to flashing ... gotta give it a bit more thought

**Comment by stdevAlDen** [comment]
So far now I was not able to read the RDP level from the firmware, only from the bootloader, it is not useful to do it in the bootloader, because if you do not enter in bootloader mode, you will not see it. An alternative to enforce the `MEMORY_PROTECTION` to be enabled is check the value at boot time and if the value is not `2` abort the boot and show a screen message informing about the problem

**Comment by stdevAlDen** [comment]
I'm trying to figure out how [`memory_protection_enabled`](https://github.com/skycoin/hardware-wallet/issues/45#issuecomment-455937763) but it's difficult if you look at the specification:
![rdp_bits](https://user-images.githubusercontent.com/34870496/56397624-ab09f280-6212-11e9-8d71-ef25cd6cd9bd.png)

so a simple check over the bits in range `15:8` in the `FLASH_OPTCR` should be enough, but it's required to think in this to avoid a mistake, maybe I'm missing something

**Comment by stdevAlDen** [comment]
I think that there is a bug here:
https://github.com/skycoin/hardware-wallet/blob/384c508d07b8909c58f9afba5b8677fc409e66bb/tiny-firmware/memory.c#L48
as this line is setting the `FLASH_OPTCR` register as a whole and maybe setting the bit in `15:8` range should be enough to avoid overwriting other flags by accident.

**Comment by stdevAlDen** [comment]
> I'm trying to figure out how [`memory_protection_enabled`](https://github.com/skycoin/hardware-wallet/issues/45#issuecomment-455937763) but it's difficult if you look at the specification:
> ![rdp_bits](https://user-images.githubusercontent.com/34870496/56397624-ab09f280-6212-11e9-8d71-ef25cd6cd9bd.png)
> 
> so a simple check over the bits in range `15:8` in the `FLASH_OPTCR` should be enough, but it's required to think in this to avoid a mistake, maybe I'm missing something

I'm having problems to read the `FLASH_OPTCR` register, the `hard_fault_handler` is getting triggered, the code is:
```c
uint8_t memory_protection_level(void) {
	uint32_t *opts_register = (uint32_t*)FLASH_OPTCR;
	uint32_t mask = 0B00000000000000001111111100000000;
	switch (((*opts_register) & mask) >> 8) {
		case 0xAA:
			return 0;
		case 0xCC:
			return 2;
		default:
			return 1;
	}
}
```

Looking at a different example where the library is st peripherals(different library but the important detai is the use of `HAL_FLASH_OB_Unlock`):
```c
void lora_flash_read_protection_check( void ) {
   /* Unlock the Flash to enable the flash control register access *************/
   HAL_FLASH_Unlock();
   /* Unlock the Options Bytes *************************************************/
   HAL_FLASH_OB_Unlock();
   FLASH_OBProgramInitTypeDef OptionsBytesStruct;
   /* Get pages write protection status ****************************************/
   HAL_FLASHEx_OBGetConfig(&OptionsBytesStruct);
   switch( OptionsBytesStruct.RDPLevel ) {
      case OB_RDP_LEVEL_0: PRINTF('RDPLevel : LEVEL 0 \r\n' ); break;
      case OB_RDP_LEVEL_1: PRINTF('RDPLevel : LEVEL 1 \r\n' ); break;
      case OB_RDP_LEVEL_2: PRINTF('RDPLevel : LEVEL 2 \r\n' ); break;
   }
   uint8_t bRDPSet = false;
   /* Check if readoutprtection is enabled ***********************/
   if((OptionsBytesStruct.RDPLevel) == OB_RDP_LEVEL_0)
   {
      OptionsBytesStruct.OptionType= OPTIONBYTE_RDP;
      OptionsBytesStruct.RDPLevel = OB_RDP_LEVEL_1;
      if(HAL_FLASHEx_OBProgram(&OptionsBytesStruct) != HAL_OK) {
         PRINTF('Fail to Read-Out Protection Settings! \r\n');
      } else
      bRDPSet = true;
      /* Generate System Reset to load the new option byte values ***************/
      HAL_FLASH_OB_Launch();
   }
   /* Lock the Options Bytes *************************************************/
   HAL_FLASH_OB_Lock();
   HAL_FLASH_Lock();
}
```

**Comment by stdevAlDen** [comment]
@olemis very important, security related
https://community.st.com/s/question/0D50X00009Xke7aSAB/readout-protection-cracked-on-stm32
https://www.aisec.fraunhofer.de/en/FirmwareProtection.html

**Comment by stdevAlDen** [comment]
> > I'm trying to figure out how [`memory_protection_enabled`](https://github.com/skycoin/hardware-wallet/issues/45#issuecomment-455937763) but it's difficult if you look at the specification:
> > ![rdp_bits](https://user-images.githubusercontent.com/34870496/56397624-ab09f280-6212-11e9-8d71-ef25cd6cd9bd.png)
> > so a simple check over the bits in range `15:8` in the `FLASH_OPTCR` should be enough, but it's required to think in this to avoid a mistake, maybe I'm missing something
> 
> I'm having problems to read the `FLASH_OPTCR` register, the `hard_fault_handler` is getting triggered, the code is:
> 
> ```c
> uint8_t memory_protection_level(void) {
> 	uint32_t *opts_register = (uint32_t*)FLASH_OPTCR;
> 	uint32_t mask = 0B00000000000000001111111100000000;
> 	switch (((*opts_register) & mask) >> 8) {
> 		case 0xAA:
> 			return 0;
> 		case 0xCC:
> 			return 2;
> 		default:
> 			return 1;
> 	}
> }
> ```
> 
> Looking at a different example where the library is st peripherals(different library but the important detai is the use of `HAL_FLASH_OB_Unlock`):
> 
> ```c
> void lora_flash_read_protection_check( void ) {
>    /* Unlock the Flash to enable the flash control register access *************/
>    HAL_FLASH_Unlock();
>    /* Unlock the Options Bytes *************************************************/
>    HAL_FLASH_OB_Unlock();
>    FLASH_OBProgramInitTypeDef OptionsBytesStruct;
>    /* Get pages write protection status ****************************************/
>    HAL_FLASHEx_OBGetConfig(&OptionsBytesStruct);
>    switch( OptionsBytesStruct.RDPLevel ) {
>       case OB_RDP_LEVEL_0: PRINTF('RDPLevel : LEVEL 0 \r\n' ); break;
>       case OB_RDP_LEVEL_1: PRINTF('RDPLevel : LEVEL 1 \r\n' ); break;
>       case OB_RDP_LEVEL_2: PRINTF('RDPLevel : LEVEL 2 \r\n' ); break;
>    }
>    uint8_t bRDPSet = false;
>    /* Check if readoutprtection is enabled ***********************/
>    if((OptionsBytesStruct.RDPLevel) == OB_RDP_LEVEL_0)
>    {
>       OptionsBytesStruct.OptionType= OPTIONBYTE_RDP;
>       OptionsBytesStruct.RDPLevel = OB_RDP_LEVEL_1;
>       if(HAL_FLASHEx_OBProgram(&OptionsBytesStruct) != HAL_OK) {
>          PRINTF('Fail to Read-Out Protection Settings! \r\n');
>       } else
>       bRDPSet = true;
>       /* Generate System Reset to load the new option byte values ***************/
>       HAL_FLASH_OB_Launch();
>    }
>    /* Lock the Options Bytes *************************************************/
>    HAL_FLASH_OB_Lock();
>    HAL_FLASH_Lock();
> }
> ```

the problem look to be related to a required call to:
```
	flash_unlock();
	flash_unlock_option_bytes();
```
but it's not working either.

**Comment by stdevAlDen** [comment]
> @olemis very important, security related
> https://community.st.com/s/question/0D50X00009Xke7aSAB/readout-protection-cracked-on-stm32
> https://www.aisec.fraunhofer.de/en/FirmwareProtection.html

@olemis https://github.com/trezor/trezor-firmware/blob/b4cec0ae541adca38ccfc46ffe6ec6ff0a3038fb/legacy/memory.c#L62

**Comment by olemis** [comment]
fixed in 95f43f7
